src/python/grpcio_tests/tests/unit/_server_wait_for_termination_test.py

```diff
@@ -0,0 +1,73 @@
+# Copyright 2019 The gRPC Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+from concurrent import futures
+import unittest
+import time
+import threading
+import six
+
+import grpc
+from tests.unit.framework.common import test_constants
+
+_WAIT_FOR_BLOCKING = datetime.timedelta(seconds=1)
+
+
+def _block_on_waiting(server, termination_event):
+    server.start()
+    server.wait_for_termination()
+    termination_event.set()
+
+
+class ServerWaitForTerminationTest(unittest.TestCase):
+
+    def test_unblock_by_invoking_stop(self):
+        termination_event = threading.Event()
+        server = grpc.server(futures.ThreadPoolExecutor())
+
+        wait_thread = threading.Thread(
+            target=_block_on_waiting, args=(
+                server,
+                termination_event,
+            ))
+        wait_thread.daemon = True
+        wait_thread.start()
+        time.sleep(_WAIT_FOR_BLOCKING.total_seconds())
+
+        server.stop(None)
+        termination_event.wait(timeout=test_constants.SHORT_TIMEOUT)
+        self.assertTrue(termination_event.is_set())
+
+    def test_unblock_by_del(self):
+        termination_event = threading.Event()
+        server = grpc.server(futures.ThreadPoolExecutor())
+
+        wait_thread = threading.Thread(
+            target=_block_on_waiting, args=(
+                server,
+                termination_event,
+            ))
+        wait_thread.daemon = True
+        wait_thread.start()
+        time.sleep(_WAIT_FOR_BLOCKING.total_seconds())
+
+        # Invoke manually here, in Python 2 it will be invoked by GC sometime.
+        server.__del__()
```

Comment:
I think `del server` is more conventional.

Reply:
`del server` is not going to trigger `__del__` stably in Python 2, and won't trigger `__del__` in Python 3 unless it is the very last ref.